### ISSUE TITLE
Python bindings setup assumes too many things

### ISSUE THREAD

**Issue by marcespie** [mlpack/mlpack#1521]
Around line 40 of setup.py.in:

                  # CMAKE_CXX_FLAGS seems to have an extra space.
                  extra_compile_args=('-DBINDING_TYPE=BINDING_TYPE_PYX ' + \
                      '-std=c++11${CMAKE_CXX_FLAGS}').split(' '),

nope, it doesn't. Not always.  Funny seeing my compiler trying to build with
-std=c++11-O2

and right after that:
                 extra_link_args=['${OpenMP_CXX_FLAGS}'],

turns into a magnificent NOT FOUND while building extensions. I have no idea (yet) whether OpenMP is required for python bindings, or whether a specific build configuration accidentally has it.

If it's required, the main CMakeLists.txt logic should then proceed to disable python extensions...

**Comment by marcespie** [comment]
I can confirm that the python bindings build just fine without ${OpenMP_CXX_FLAGS} on a !OpenMP installation...

**Comment by rcurtin** [comment]
Ah, thanks.  I think that the two issues you've encountered here are related to #1478 and #1425, respectively.  In both of those cases, it was problematic to come up with fixes because we couldn't directly reproduce them.

Can you try the patch in #1478 and see if that works for you?  If so, then I'd be happy to commit that.

And for the OpenMP CXXFLAGS bit, I bet this is what was going wrong in #1425, so thanks for pointing it out.  We can probably do the following... in `CMakeLists.txt`, change the block at line 360 from:

```
if (OPENMP_FOUND)
  add_definitions(-DHAS_OPENMP)
  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${OpenMP_C_FLAGS}")
  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS}")
else ()
  # Disable warnings for all the unknown OpenMP pragmas.
  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-unknown-pragmas")
endif ()
```

to

```
if (OPENMP_FOUND)
  add_definitions(-DHAS_OPENMP)
  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${OpenMP_C_FLAGS}")
  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS}")
else ()
  # Disable warnings for all the unknown OpenMP pragmas.
  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-unknown-pragmas")
  set(OpenMP_CXX_FLAGS "")
endif ()
```

If that works, I'll open a PR with these two fixes and also the ones from the other issues you've opened.  Thanks again for the reports, these are all great things to get fixed. :+1:

**Comment by rcurtin** [comment]
Fixed in #1523.
